cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/remote/RemoteExecutionTwillRunnerService.java

```diff
@@ -514,200 +502,59 @@ public TwillController createTwillControllerFromRunRecord(RunRecordDetail runRec
-      Map<String, String> systemArguments = programOpts.getArguments().asMap();
-      String provisionerName = SystemArguments.getProfileProvisioner(systemArguments);
-      String peerName = systemArguments.get(ProgramOptionConstants.PEER_NAME);
-      boolean useControllerToStop = TetheringProvisioner.TETHERING_NAME.equals(provisionerName) &&
-        peerName == null;
-      // Create the controller and start the runtime monitor when the startup task completed successfully.
-      RemoteExecutionTwillController controller = new RemoteExecutionTwillController(cConf, programRunId,
-                                                                                     startupTaskCompletion,
-                                                                                     processController,
-                                                                                     scheduler, remoteExecutionService,
-                                                                                     useControllerToStop);
-      startupTaskCompletion.thenAccept(o -> remoteExecutionService.start());
-
-      // On this controller termination, make sure it is removed from the controllers map and have resources released.
-      controller.onTerminated(() -> {
-        if (controllers.remove(programRunId, controller)) {
-          controller.complete();
-        }
-      }, scheduler);
-      return controller;
-    }
+  ProgramStateWriter getProgramStateWriter() {
+    return programStateWriter;
+  }
 
-    private RemoteProcessController createRemoteProcessController(ProgramRunId programRunId,
-                                                                  ProgramOptions programOpts) {
-      RuntimeJobManager jobManager = provisioningService.getRuntimeJobManager(programRunId, programOpts).orElse(null);
-      // Use RuntimeJobManager to control the remote process if it is supported
-      if (jobManager != null) {
-        LOG.debug("Creating controller for program run {} with runtime job manager", programRunId);
-        return new RuntimeJobRemoteProcessController(
-          programRunId,
-          () -> provisioningService.getRuntimeJobManager(programRunId, programOpts)
-            .orElseThrow(IllegalStateException::new));
-      }
+  ServiceSocksProxy getServiceSocksProxy() {
+    return serviceSocksProxy;
+  }
 
-      // Otherwise, default to SSH
-      ClusterKeyInfo clusterKeyInfo = new ClusterKeyInfo(cConf, programOpts, locationFactory);
-      SSHConfig sshConfig = clusterKeyInfo.getSSHConfig();
-      LOG.debug("Creating controller for program run {} with SSH config {}", programRunId, sshConfig);
+  RuntimeServiceSocksProxyAuthenticator getServiceSocksProxyAuthenticator() {
+    return serviceSocksProxyAuthenticator;
+  }
 
-      return new SSHRemoteProcessController(programRunId, programOpts, sshConfig, provisioningService);
-    }
+  Lock getControllersLock() {
```

Comment:
adding getters to expose all these internals isn't so great as it breaks the encapsulation. Looks like this is only done to make them available to the ControllerFactory. Seems like the FireAndForgetControllerFactory only needs the cconf and scheduler but not all this other stuff.

Would it work if we keep the ControllerFactory within this class, and instead refactor things so that the cconf and scheduler get passed down to the necessary method that FireAndForget overrides?

Reply:
No need for these getters anymore.